kms/tpmkms/tpmkms.go

```diff
@@ -0,0 +1,674 @@
+			Size:           size,
+			QualifyingData: properties.qualifyingData,
+		}
+		key, err = k.tpm.AttestKey(ctx, properties.attestBy, properties.name, config)
+		if err != nil {
+			if errors.Is(err, tpm.ErrExists) {
+				return nil, apiv1.AlreadyExistsError{Message: err.Error()}
+			}
+			return nil, fmt.Errorf("failed creating attested key: %w", err)
+		}
+	} else {
+		config := tpm.CreateKeyConfig{
+			Algorithm: v.Type,
+			Size:      size,
+		}
+		key, err = k.tpm.CreateKey(ctx, properties.name, config)
+		if err != nil {
+			if errors.Is(err, tpm.ErrExists) {
+				return nil, apiv1.AlreadyExistsError{Message: err.Error()}
+			}
+			return nil, fmt.Errorf("failed creating key: %w", err)
+		}
+	}
+
+	signer, err := key.Signer(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed getting signer for key: %w", err)
+	}
+
+	createdKeyURI := fmt.Sprintf("tpmkms:name=%s", key.Name())
+	if properties.attestBy != "" {
+		createdKeyURI = fmt.Sprintf("%s;attest-by=%s", createdKeyURI, key.AttestedBy())
+	}
+
+	return &apiv1.CreateKeyResponse{
+		Name:      createdKeyURI,
+		PublicKey: signer.Public(),
+		CreateSignerRequest: apiv1.CreateSignerRequest{
+			SigningKey: createdKeyURI,
+			Signer:     signer,
+		},
+	}, nil
+}
+
+// CreateSigner creates a signer using a key present in the TPM KMS.
+func (k *TPMKMS) CreateSigner(req *apiv1.CreateSignerRequest) (crypto.Signer, error) {
+	if req.Signer != nil {
+		return req.Signer, nil
+	}
+
+	if req.SigningKey == "" {
+		return nil, errors.New("createSignerRequest 'signingKey' cannot be empty")
+	}
+
+	properties, err := parseNameURI(req.SigningKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed parsing %q: %w", req.SigningKey, err)
+	}
+
+	if properties.ak {
+		return nil, fmt.Errorf("signing with an AK currently not supported")
+	}
+
+	ctx := context.Background()
+	key, err := k.tpm.GetKey(ctx, properties.name)
+	if err != nil {
+		return nil, err
+	}
+
+	signer, err := key.Signer(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed getting signer for key %q: %w", properties.name, err)
+	}
+
+	return signer, nil
+}
+
+// GetPublicKey returns the public key ....
+func (k *TPMKMS) GetPublicKey(req *apiv1.GetPublicKeyRequest) (crypto.PublicKey, error) {
+	if req.Name == "" {
+		return nil, errors.New("getPublicKeyRequest 'name' cannot be empty")
+	}
+
+	properties, err := parseNameURI(req.Name)
+	if err != nil {
+		return nil, fmt.Errorf("failed parsing %q: %w", req.Name, err)
+	}
+
+	if properties.ak {
+		return nil, fmt.Errorf("retrieving AK public key currently not supported")
+	}
+
+	ctx := context.Background()
+	key, err := k.tpm.GetKey(ctx, properties.name)
+	if err != nil {
+		return nil, err
+	}
+
+	signer, err := key.Signer(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed getting signer for key %q: %w", properties.name, err)
+	}
+
+	return signer.Public(), nil
+}
+
+func (k *TPMKMS) LoadCertificate(req *apiv1.LoadCertificateRequest) (*x509.Certificate, error) {
```

Comment:
This method could call `LoadCertificateChain` and return the leaf.

Reply:
Done in dc0a70717a587b95079a181a630c83126f150c2b.